#Задание

Используя шаблон ноутбука для распознавания видов одежды и аксессуаров из набора **fashion_mnist**, выполните следующие действия:

1. Создайте **9** моделей нейронной сети с различными архитектурами и сравните в них значения точности на проверочной выборке (на последней эпохе) и на тестовой выборке.  Используйте следующее деление: обучающая выборка - **50000** примеров, проверочная выборка - **10000** примеров, тестовая выборка - **10000** примеров.

2. Создайте сравнительную таблицу в конце ноутбука, напишите свои выводы по результатам проведенных тестов.

# Шаблон ноутбука

##Импорт библиотек

In [ ]:
# Последовательная модель НС
from tensorflow.keras.models import Sequential

# Основные слои
from tensorflow.keras.layers import Dense, Activation, Dropout, BatchNormalization

# Утилиты для to_categorical()
from tensorflow.keras import utils

# Алгоритмы оптимизации для обучения модели
from tensorflow.keras.optimizers import Adam, Adadelta

# Библиотека для работы с массивами
import numpy as np

# Библиотека для работы с таблицами
import pandas as pd

# Отрисовка графиков
import matplotlib.pyplot as plt

# Связь с google-диском
from google.colab import files

# Предварительная обработка данных
from sklearn import preprocessing

# Разделение данных на выборки
from sklearn.model_selection import train_test_split

# Для загрузки датасета
from keras.datasets import fashion_mnist

# Отрисовывать изображения в ноутбуке, а не в консоль или файл
%matplotlib inline

##Описание базы

###База: одежда, обувь и аксессуары
- Датасет состоит из набора изображений одежды, обуви, аксессуаров и их классов.
- Изображения одного вида хранятся в numpy-массиве **(28, 28)** - **x_train, x_test**.
- База содержит **10** классов: (Футболка, Брюки, Пуловер, Платье, Пальто, Сандалии/Босоножки, Рубашка, Кроссовки, Сумочка, Ботильоны) - **y_train, y_test**.
- Примеров: train - **60000**, test - **10000**.

###Вывод примеров

In [ ]:
# Загрузка датасета
(x_train_org, y_train_org), (x_test, y_test) = fashion_mnist.load_data()

# Вывод размерностей выборок
x_train, x_val = x_train_org[:50000], x_train_org[50000:]
y_train, y_val = y_train_org[:50000], y_train_org[50000:]

print(x_train.shape)
print(x_val.shape)
print(x_test.shape)

29515/29515 ━━━━━━━━━━━━━━━━━━━━ 0s 0us/step
26421880/26421880 ━━━━━━━━━━━━━━━━━━━━ 0s 0us/step
5148/5148 ━━━━━━━━━━━━━━━━━━━━ 0s 0us/step
4422102/4422102 ━━━━━━━━━━━━━━━━━━━━ 0s 0us/step
(50000, 28, 28)
(10000, 28, 28)
(10000, 28, 28)


In [ ]:
x_train = x_train.astype('float32') / 255.
x_val = x_val.astype('float32') / 255.
x_test = x_test.astype('float32') / 255.

In [ ]:
x_train = x_train.reshape(x_train.shape[0], -1)
x_val   = x_val.reshape(x_val.shape[0], -1)
x_test  = x_test.reshape(x_test.shape[0], -1)


In [ ]:
CLASS_COUNT = 10

In [ ]:
y_train = utils.to_categorical(y_train, CLASS_COUNT)
y_val = utils.to_categorical(y_val, CLASS_COUNT)
y_test = utils.to_categorical(y_test, CLASS_COUNT)

In [ ]:
print(x_train.shape)
print(y_train.shape)
print(x_val.shape)
print(y_val.shape)

(50000, 784)
(50000, 10)
(10000, 784)
(10000, 10)


In [ ]:
def create_model(units1, units2, activation1, activation2):
    model = Sequential()

    model.add(Dense(units1, input_dim=x_train.shape[1], activation=activation1))

    model.add(Dense(units2, activation=activation2))
    model.add(Dense(CLASS_COUNT, activation='softmax'))

    model.compile(
        loss='categorical_crossentropy',
        optimizer='adam',
        metrics=['accuracy']
    )

    return model

In [ ]:
val_accuracies = {}
test_accuracies = {}

In [ ]:
model1 = create_model(60, 30, 'relu', 'relu')
history1 = model1.fit(x_train,        # обучающая выборка, входные данные
          y_train,        # обучающая выборка, выходные данные
          batch_size=64,# кол-во примеров, которое обрабатывает нейронка перед одним изменением весов
          validation_data=(x_val, y_val),
          epochs=5,      # количество эпох, когда нейронка обучается на всех примерах выборки
          verbose=1)

val_accuracies['model1'] = history1.history['val_accuracy'][-1]  # последняя эпоха
test_loss, test_acc = model1.evaluate(x_test, y_test, verbose=0)
print()
test_accuracies['model1'] = test_acc
print("Test accuracy:", test_accuracies['model1'])

/usr/local/lib/python3.12/dist-packages/keras/src/layers/core/dense.py:106: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


Epoch 1/5
782/782 ━━━━━━━━━━━━━━━━━━━━ 6s 5ms/step - accuracy: 0.8055 - loss: 0.5650 - val_accuracy: 0.8375 - val_loss: 0.4616
Epoch 2/5
782/782 ━━━━━━━━━━━━━━━━━━━━ 2s 3ms/step - accuracy: 0.8540 - loss: 0.4131 - val_accuracy: 0.8552 - val_loss: 0.4122
Epoch 3/5
782/782 ━━━━━━━━━━━━━━━━━━━━ 3s 4ms/step - accuracy: 0.8683 - loss: 0.3683 - val_accuracy: 0.8631 - val_loss: 0.3931
Epoch 4/5
782/782 ━━━━━━━━━━━━━━━━━━━━ 2s 3ms/step - accuracy: 0.8753 - loss: 0.3458 - val_accuracy: 0.8726 - val_loss: 0.3636
Epoch 5/5
782/782 ━━━━━━━━━━━━━━━━━━━━ 2s 3ms/step - accuracy: 0.8808 - loss: 0.3282 - val_accuracy: 0.8686 - val_loss: 0.3667

Test accuracy: 0.8626999855041504


In [ ]:
model2 = create_model(128, 64, 'relu', 'relu')
history2 = model2.fit(x_train,        # обучающая выборка, входные данные
          y_train,        # обучающая выборка, выходные данные
          batch_size=128,# кол-во примеров, которое обрабатывает нейронка перед одним изменением весов
          validation_data=(x_val, y_val),
          epochs=5,      # количество эпох, когда нейронка обучается на всех примерах выборки
          verbose=1)

val_accuracies['model2'] = history2.history['val_accuracy'][-1]  # последняя эпоха
test_loss, test_acc = model2.evaluate(x_test, y_test, verbose=0)
print()
test_accuracies['model2'] = test_acc
print("Test accuracy:", test_accuracies['model2'])

/usr/local/lib/python3.12/dist-packages/keras/src/layers/core/dense.py:106: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


Epoch 1/5
391/391 ━━━━━━━━━━━━━━━━━━━━ 4s 7ms/step - accuracy: 0.7996 - loss: 0.5803 - val_accuracy: 0.8459 - val_loss: 0.4358
Epoch 2/5
391/391 ━━━━━━━━━━━━━━━━━━━━ 1s 3ms/step - accuracy: 0.8554 - loss: 0.4053 - val_accuracy: 0.8540 - val_loss: 0.4050
Epoch 3/5
391/391 ━━━━━━━━━━━━━━━━━━━━ 1s 3ms/step - accuracy: 0.8713 - loss: 0.3579 - val_accuracy: 0.8693 - val_loss: 0.3622
Epoch 4/5
391/391 ━━━━━━━━━━━━━━━━━━━━ 1s 3ms/step - accuracy: 0.8816 - loss: 0.3291 - val_accuracy: 0.8787 - val_loss: 0.3476
Epoch 5/5
391/391 ━━━━━━━━━━━━━━━━━━━━ 2s 4ms/step - accuracy: 0.8848 - loss: 0.3128 - val_accuracy: 0.8770 - val_loss: 0.3512

Test accuracy: 0.8682000041007996


In [ ]:
model3 = create_model(256, 128, 'tanh', 'relu')
history3 = model3.fit(x_train,        # обучающая выборка, входные данные
          y_train,        # обучающая выборка, выходные данные
          batch_size=32,# кол-во примеров, которое обрабатывает нейронка перед одним изменением весов
          validation_data=(x_val, y_val),
          epochs=5,      # количество эпох, когда нейронка обучается на всех примерах выборки
          verbose=1)

val_accuracies['model3'] = history3.history['val_accuracy'][-1]  # последняя эпоха
test_loss, test_acc = model3.evaluate(x_test, y_test, verbose=0)
print()
test_accuracies['model3'] = test_acc
print("Test accuracy:", test_accuracies['model3'])

Epoch 1/5
1563/1563 ━━━━━━━━━━━━━━━━━━━━ 8s 4ms/step - accuracy: 0.8290 - loss: 0.4730 - val_accuracy: 0.8336 - val_loss: 0.4549
Epoch 2/5
1563/1563 ━━━━━━━━━━━━━━━━━━━━ 4s 3ms/step - accuracy: 0.8625 - loss: 0.3691 - val_accuracy: 0.8699 - val_loss: 0.3557
Epoch 3/5
1563/1563 ━━━━━━━━━━━━━━━━━━━━ 4s 3ms/step - accuracy: 0.8752 - loss: 0.3319 - val_accuracy: 0.8698 - val_loss: 0.3507
Epoch 4/5
1563/1563 ━━━━━━━━━━━━━━━━━━━━ 5s 3ms/step - accuracy: 0.8828 - loss: 0.3099 - val_accuracy: 0.8771 - val_loss: 0.3343
Epoch 5/5
1563/1563 ━━━━━━━━━━━━━━━━━━━━ 4s 3ms/step - accuracy: 0.8900 - loss: 0.2931 - val_accuracy: 0.8673 - val_loss: 0.3573

Test accuracy: 0.8592000007629395


In [ ]:
model4 = create_model(256, 64, 'tanh', 'tanh')
history4 = model4.fit(x_train,        # обучающая выборка, входные данные
          y_train,        # обучающая выборка, выходные данные
          batch_size=64,# кол-во примеров, которое обрабатывает нейронка перед одним изменением весов
          validation_data=(x_val, y_val),
          epochs=5,      # количество эпох, когда нейронка обучается на всех примерах выборки
          verbose=1)

val_accuracies['model4'] = history4.history['val_accuracy'][-1]  # последняя эпоха
test_loss, test_acc = model4.evaluate(x_test, y_test, verbose=0)
print()
test_accuracies['model4'] = test_acc
print("Test accuracy:", test_accuracies['model4'])

Epoch 1/5
782/782 ━━━━━━━━━━━━━━━━━━━━ 5s 4ms/step - accuracy: 0.8225 - loss: 0.4910 - val_accuracy: 0.8454 - val_loss: 0.4116
Epoch 2/5
782/782 ━━━━━━━━━━━━━━━━━━━━ 2s 3ms/step - accuracy: 0.8645 - loss: 0.3735 - val_accuracy: 0.8625 - val_loss: 0.3715
Epoch 3/5
782/782 ━━━━━━━━━━━━━━━━━━━━ 2s 3ms/step - accuracy: 0.8741 - loss: 0.3410 - val_accuracy: 0.8600 - val_loss: 0.3765
Epoch 4/5
782/782 ━━━━━━━━━━━━━━━━━━━━ 2s 3ms/step - accuracy: 0.8842 - loss: 0.3168 - val_accuracy: 0.8721 - val_loss: 0.3566
Epoch 5/5
782/782 ━━━━━━━━━━━━━━━━━━━━ 2s 3ms/step - accuracy: 0.8916 - loss: 0.2954 - val_accuracy: 0.8783 - val_loss: 0.3359

Test accuracy: 0.8686000108718872


In [ ]:
model5 = create_model(64, 32, 'sigmoid', 'relu')
history5 = model5.fit(x_train,        # обучающая выборка, входные данные
          y_train,        # обучающая выборка, выходные данные
          batch_size=64,# кол-во примеров, которое обрабатывает нейронка перед одним изменением весов
          validation_data=(x_val, y_val),
          epochs=5,      # количество эпох, когда нейронка обучается на всех примерах выборки
          verbose=1)

val_accuracies['model5'] = history5.history['val_accuracy'][-1]  # последняя эпоха
test_loss, test_acc = model5.evaluate(x_test, y_test, verbose=0)
print()
test_accuracies['model5'] = test_acc
print("Test accuracy:", test_accuracies['model5'])

Epoch 1/5
782/782 ━━━━━━━━━━━━━━━━━━━━ 6s 5ms/step - accuracy: 0.7785 - loss: 0.6712 - val_accuracy: 0.8355 - val_loss: 0.4530
Epoch 2/5
782/782 ━━━━━━━━━━━━━━━━━━━━ 3s 3ms/step - accuracy: 0.8501 - loss: 0.4189 - val_accuracy: 0.8548 - val_loss: 0.3960
Epoch 3/5
782/782 ━━━━━━━━━━━━━━━━━━━━ 2s 3ms/step - accuracy: 0.8640 - loss: 0.3771 - val_accuracy: 0.8644 - val_loss: 0.3706
Epoch 4/5
782/782 ━━━━━━━━━━━━━━━━━━━━ 2s 3ms/step - accuracy: 0.8739 - loss: 0.3502 - val_accuracy: 0.8672 - val_loss: 0.3555
Epoch 5/5
782/782 ━━━━━━━━━━━━━━━━━━━━ 2s 3ms/step - accuracy: 0.8783 - loss: 0.3331 - val_accuracy: 0.8731 - val_loss: 0.3433

Test accuracy: 0.8640999794006348


In [ ]:
model6 = create_model(128, 32, 'relu', 'tanh')
history6 = model6.fit(x_train,        # обучающая выборка, входные данные
          y_train,        # обучающая выборка, выходные данные
          batch_size=512,# кол-во примеров, которое обрабатывает нейронка перед одним изменением весов
          validation_data=(x_val, y_val),
          epochs=5,      # количество эпох, когда нейронка обучается на всех примерах выборки
          verbose=1)

val_accuracies['model6'] = history6.history['val_accuracy'][-1]  # последняя эпоха
test_loss, test_acc = model6.evaluate(x_test, y_test, verbose=0)
print()
test_accuracies['model6'] = test_acc
print("Test accuracy:", test_accuracies['model6'])

Epoch 1/5
98/98 ━━━━━━━━━━━━━━━━━━━━ 4s 20ms/step - accuracy: 0.7418 - loss: 0.7960 - val_accuracy: 0.8122 - val_loss: 0.5365
Epoch 2/5
98/98 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.8399 - loss: 0.4692 - val_accuracy: 0.8482 - val_loss: 0.4382
Epoch 3/5
98/98 ━━━━━━━━━━━━━━━━━━━━ 1s 4ms/step - accuracy: 0.8575 - loss: 0.4078 - val_accuracy: 0.8623 - val_loss: 0.4023
Epoch 4/5
98/98 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.8677 - loss: 0.3757 - val_accuracy: 0.8583 - val_loss: 0.3881
Epoch 5/5
98/98 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.8724 - loss: 0.3557 - val_accuracy: 0.8680 - val_loss: 0.3668

Test accuracy: 0.8629000186920166


In [ ]:
model7 = create_model(32, 16, 'relu', 'tanh')
history7= model7.fit(x_train,        # обучающая выборка, входные данные
          y_train,        # обучающая выборка, выходные данные
          batch_size=64,# кол-во примеров, которое обрабатывает нейронка перед одним изменением весов
          validation_data=(x_val, y_val),
          epochs=5,      # количество эпох, когда нейронка обучается на всех примерах выборки
          verbose=1)

val_accuracies['model7'] = history7.history['val_accuracy'][-1]  # последняя эпоха
test_loss, test_acc = model7.evaluate(x_test, y_test, verbose=0)
print()
test_accuracies['model7'] = test_acc
print("Test accuracy:", test_accuracies['model7'])

Epoch 1/5
782/782 ━━━━━━━━━━━━━━━━━━━━ 5s 5ms/step - accuracy: 0.7857 - loss: 0.7048 - val_accuracy: 0.8331 - val_loss: 0.4837
Epoch 2/5
782/782 ━━━━━━━━━━━━━━━━━━━━ 2s 3ms/step - accuracy: 0.8478 - loss: 0.4377 - val_accuracy: 0.8514 - val_loss: 0.4185
Epoch 3/5
782/782 ━━━━━━━━━━━━━━━━━━━━ 2s 3ms/step - accuracy: 0.8618 - loss: 0.3932 - val_accuracy: 0.8632 - val_loss: 0.3885
Epoch 4/5
782/782 ━━━━━━━━━━━━━━━━━━━━ 3s 3ms/step - accuracy: 0.8687 - loss: 0.3687 - val_accuracy: 0.8558 - val_loss: 0.3970
Epoch 5/5
782/782 ━━━━━━━━━━━━━━━━━━━━ 2s 3ms/step - accuracy: 0.8732 - loss: 0.3538 - val_accuracy: 0.8625 - val_loss: 0.3935

Test accuracy: 0.8539999723434448


In [ ]:
model8 = create_model(64, 32, 'tanh', 'tanh')
history8 = model8.fit(x_train,        # обучающая выборка, входные данные
          y_train,        # обучающая выборка, выходные данные
          batch_size=32,# кол-во примеров, которое обрабатывает нейронка перед одним изменением весов
          validation_data=(x_val, y_val),
          epochs=5,      # количество эпох, когда нейронка обучается на всех примерах выборки
          verbose=1)

val_accuracies['model8'] = history8.history['val_accuracy'][-1]  # последняя эпоха
test_loss, test_acc = model8.evaluate(x_test, y_test, verbose=0)
print()
test_accuracies['model8'] = test_acc
print("Test accuracy:", test_accuracies['model8'])

Epoch 1/5
1563/1563 ━━━━━━━━━━━━━━━━━━━━ 8s 4ms/step - accuracy: 0.8179 - loss: 0.5188 - val_accuracy: 0.8456 - val_loss: 0.4223
Epoch 2/5
1563/1563 ━━━━━━━━━━━━━━━━━━━━ 4s 3ms/step - accuracy: 0.8637 - loss: 0.3818 - val_accuracy: 0.8591 - val_loss: 0.3842
Epoch 3/5
1563/1563 ━━━━━━━━━━━━━━━━━━━━ 5s 3ms/step - accuracy: 0.8722 - loss: 0.3484 - val_accuracy: 0.8584 - val_loss: 0.3777
Epoch 4/5
1563/1563 ━━━━━━━━━━━━━━━━━━━━ 4s 3ms/step - accuracy: 0.8796 - loss: 0.3256 - val_accuracy: 0.8764 - val_loss: 0.3388
Epoch 5/5
1563/1563 ━━━━━━━━━━━━━━━━━━━━ 4s 3ms/step - accuracy: 0.8860 - loss: 0.3096 - val_accuracy: 0.8749 - val_loss: 0.3432

Test accuracy: 0.8659999966621399


In [ ]:
model9 = create_model(256, 128, 'relu', 'sigmoid')
history9 = model9.fit(x_train,        # обучающая выборка, входные данные
          y_train,        # обучающая выборка, выходные данные
          batch_size=64,# кол-во примеров, которое обрабатывает нейронка перед одним изменением весов
          validation_data=(x_val, y_val),
          epochs=5,      # количество эпох, когда нейронка обучается на всех примерах выборки
          verbose=1)

val_accuracies['model9'] = history9.history['val_accuracy'][-1]  # последняя эпоха
test_loss, test_acc = model9.evaluate(x_test, y_test, verbose=0)
print()
test_accuracies['model9'] = test_acc
print("Test accuracy:", test_accuracies['model9'])

Epoch 1/5
782/782 ━━━━━━━━━━━━━━━━━━━━ 5s 5ms/step - accuracy: 0.8124 - loss: 0.5400 - val_accuracy: 0.8393 - val_loss: 0.4320
Epoch 2/5
782/782 ━━━━━━━━━━━━━━━━━━━━ 2s 3ms/step - accuracy: 0.8641 - loss: 0.3728 - val_accuracy: 0.8592 - val_loss: 0.3831
Epoch 3/5
782/782 ━━━━━━━━━━━━━━━━━━━━ 2s 3ms/step - accuracy: 0.8779 - loss: 0.3344 - val_accuracy: 0.8728 - val_loss: 0.3454
Epoch 4/5
782/782 ━━━━━━━━━━━━━━━━━━━━ 2s 3ms/step - accuracy: 0.8853 - loss: 0.3071 - val_accuracy: 0.8691 - val_loss: 0.3444
Epoch 5/5
782/782 ━━━━━━━━━━━━━━━━━━━━ 2s 3ms/step - accuracy: 0.8937 - loss: 0.2873 - val_accuracy: 0.8817 - val_loss: 0.3222

Test accuracy: 0.8758999705314636


In [ ]:
import pandas as pd

df = pd.DataFrame({
    'Model': list(val_accuracies.keys()),
    'Validation Accuracy': list(val_accuracies.values()),
    'Test Accuracy': [test_accuracies[m] for m in val_accuracies.keys()]
})

df

,Model,Validation Accuracy,Test Accuracy
0,model1,0.8686,0.8627
1,model2,0.8770,0.8682
2,model3,0.8673,0.8592
3,model4,0.8783,0.8686
4,model5,0.8731,0.8641
5,model6,0.8680,0.8629
6,model7,0.8625,0.8540
7,model8,0.8749,0.8660
8,model9,0.8817,0.8759


Все 9 моделей показали стабильную и достаточно высокую точность на проверочной и тестовой выборках — от примерно 85% до 88%. Разница между точностью на проверке и на тесте очень мала, что свидетельствует о хорошей способности моделей к обобщению и отсутствии  переобучения. Увеличение количества нейронов в скрытых слоях заметно улучшает точность: модели с большими слоями, такие как model9 (256 и 128 нейронов), показали наилучшие результаты, тогда как маленькие модели, например model7, ограничены в выражающей способности и дают чуть меньшую точность.

Функции активации тоже влияют на результаты. ReLU остаётся наиболее эффективной функцией для скрытых слоев, а комбинации ReLU с другими функциями, например Sigmoid в model9, дают небольшой прирост точности. Модели с только tanh или sigmoid показывают немного меньшую точность и более медленное обучение.

Батч-сайз оказывает влияние преимущественно на стабильность и скорость обучения. Малые батчи (32–64) дают чуть лучшее обобщение за счёт более «шумных» градиентов, а большие батчи ускоряют обучение, но иногда чуть снижают точность на тестовой выборке.

Лучшие результаты показала модель model9, которая сочетает крупные скрытые слои и комбинацию функций активации. В целом, для задачи распознавания изображений Fashion MNIST оптимальными оказались модели со средними или большими скрытыми слоями и ReLU в первом слое